[ONNX] Expose OnnxRegistry publicly

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106140

The official move of `OnnxRegistry` to `torch.onnx` allows it to become one of the parameters in `torch.onnx.ExportOption`. By incorporating `OnnxRegistry` in `torch.onnx.ExportOption`, users gain access to various functionalities, including the ability to register custom operators using `register_custom_op`, check whether an operator is supported using `is_registered_op`, and obtain symbolic functions that support specific operators using `get_functions`.

Additionally, `opset_version` is now exclusively available in `torch.onnx.OnnxRegistry` as it is removed from `torch.onnx.ExportOption`. The initialization of the registry with torchlib under the provided opset version ensures that the exporter uses the specified opset version as the primary version for exporting.

These changes encompass scenarios where users can:

1. Register an unsupported ATen operator with a custom implementation using onnx-script.
2. Override an existing symbolic function (onnx invariant).

NOTE: The custom registered function will be prioritized in onnx dispatcher, and if there are multiple custom ones, the one registered the last will be picked.

titaiwangms committed:
[ONNX] Expose OnnxRegistry publicly

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Expose OnnxRegistry publicly"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Expose OnnxRegistry publicly"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Expose OnnxRegistry publicly"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Expose OnnxRegistry publicly"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Expose OnnxRegistry publicly"




[ghstack-poisoned]

titaiwangms commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -209,15 +209,29 @@ def forward(self, input):
+
+        # _DEFAULT_OPSET_VERSION = 18
+        onnx_registry = torch.onnx.OnnxRegistry()
+        self.assertTrue(
```
@BowenBao @justinchuby Why is this False? It works at local...

justinchuby replied:
I would avoid reliance on init behavior here. Potentially register an op to it, then check existence. In the future I suggest we don't populate the registry with `OnnxRegistry()` init alone.

BowenBao replied:
onnxscript version diff I suppose?

titaiwangms replied:
> onnxscript version diff I suppose?

I see. That's it.

titaiwangms replied:
> I would avoid reliance on init behavior here. Potentially register an op to it, then check existence. In the future I suggest we don't populate the registry with `OnnxRegistry()` init alone.

Could you share more of the thoughts? Creating OnnRegistry() which is initiated with full torchlib is kind of the core idea of making registry public. What is the other ways around?

BowenBao replied:
@justinchuby do you mean lazy population of torchlib registry?

🤔 I think it can work, bottom line is we hide the implementation detail and only lazy populate when needed (read) so the behavior stays the same. 

But I'm not sure what are we gaining? We can make torchlib registry read only too.

justinchuby replied:
I just think for this test to be robust it would be nice to not assume the existence of an op from torchlib.

BowenBao replied:
okay :)

titaiwangms replied:
Agree, I was going to use pure custom ops from torch, but torch only supports custom op with namespace and op_name, so I can't have it with overloads. I already put TODO on that.

titaiwangms committed:
Update on "[ONNX] Expose OnnxRegistry publicly"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Expose OnnxRegistry publicly"


The official move of `OnnxRegistry` to `torch.onnx` allows it to become one of the parameters in `torch.onnx.ExportOption`. By incorporating `OnnxRegistry` in `torch.onnx.ExportOption`, users gain access to various functionalities, including the ability to register custom operators using `register_custom_op`, check whether an operator is supported using `is_registered_op`, and obtain symbolic functions that support specific operators using `get_functions`.

Additionally, `opset_version` is now exclusively available in `torch.onnx.OnnxRegistry` as it is removed from `torch.onnx.ExportOption`. The initialization of the registry with torchlib under the provided opset version ensures that the exporter uses the specified opset version as the primary version for exporting.

These changes encompass scenarios where users can:

1. Register an unsupported ATen operator with a custom implementation using onnx-script.
2. Override an existing symbolic function (onnx invariant).

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Expose OnnxRegistry publicly"


The official move of `OnnxRegistry` to `torch.onnx` allows it to become one of the parameters in `torch.onnx.ExportOption`. By incorporating `OnnxRegistry` in `torch.onnx.ExportOption`, users gain access to various functionalities, including the ability to register custom operators using `register_custom_op`, check whether an operator is supported using `is_registered_op`, and obtain symbolic functions that support specific operators using `get_functions`.

Additionally, `opset_version` is now exclusively available in `torch.onnx.OnnxRegistry` as it is removed from `torch.onnx.ExportOption`. The initialization of the registry with torchlib under the provided opset version ensures that the exporter uses the specified opset version as the primary version for exporting.

These changes encompass scenarios where users can:

1. Register an unsupported ATen operator with a custom implementation using onnx-script.
2. Override an existing symbolic function (onnx invariant).

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/exporter.py:
```diff
@@ -75,14 +82,185 @@ class ONNXFakeContext:
+
+    """
+
+    def __init__(self, opset_version: int = _DEFAULT_OPSET_VERSION) -> None:
```
Don't expose `opset_version`

titaiwangms commented on torch/onnx/_internal/exporter.py:
```diff
@@ -75,14 +82,185 @@ class ONNXFakeContext:
+        self._registry[internal_qualified_name].append(symbolic_function)
+
+    @_beartype.beartype
+    def register_custom_op(
```
Do not allow other opset version for now?

titaiwangms committed:
Update on "[ONNX] Expose OnnxRegistry publicly"


The official move of `OnnxRegistry` to `torch.onnx` allows it to become one of the parameters in `torch.onnx.ExportOption`. By incorporating `OnnxRegistry` in `torch.onnx.ExportOption`, users gain access to various functionalities, including the ability to register custom operators using `register_custom_op`, check whether an operator is supported using `is_registered_op`, and obtain symbolic functions that support specific operators using `get_functions`.

Additionally, `opset_version` is now exclusively available in `torch.onnx.OnnxRegistry` as it is removed from `torch.onnx.ExportOption`. The initialization of the registry with torchlib under the provided opset version ensures that the exporter uses the specified opset version as the primary version for exporting.

These changes encompass scenarios where users can:

1. Register an unsupported ATen operator with a custom implementation using onnx-script.
2. Override an existing symbolic function (onnx invariant).

NOTE: The custom registered function will be prioritized in onnx dispatcher, and if there are multiple custom ones, the one registered the last will be picked.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -75,14 +82,179 @@ class ONNXFakeContext:
+        )
+
+        # opset_version is unused for now, since torchlib only supports opset18.
+        self._opset_version = _DEFAULT_OPSET_VERSION
```
nit: can we get this from onnxscript torchlib?

justinchuby replied:
We don’t have that info but we can if needed

thiagocrepaldi replied:
+1
Not having to modify the exporter (different repo) each time ONNX Script changes the default opset version is indeed a good feature. ONNX Script and exporter should be able to evolve independently.

titaiwangms replied:
It's a good idea to have the version from torchlib, but it doesn't officially have it now. I think we can keep this as it's not public, and wait until we are working on versioning, then we can figure out a more thorough way to connect converter registry and torchlib registry.

titaiwangms replied:
I will put a TODO

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -45,7 +49,10 @@
+        registration as torchlib_registry,
+    )
 
 _DEFAULT_OPSET_VERSION: Final[int] = 18
```
Following on my last comment. Is it possible to delete this if it is not used anywhere else?

titaiwangms committed:
Update on "[ONNX] Expose OnnxRegistry publicly"


The official move of `OnnxRegistry` to `torch.onnx` allows it to become one of the parameters in `torch.onnx.ExportOption`. By incorporating `OnnxRegistry` in `torch.onnx.ExportOption`, users gain access to various functionalities, including the ability to register custom operators using `register_custom_op`, check whether an operator is supported using `is_registered_op`, and obtain symbolic functions that support specific operators using `get_functions`.

Additionally, `opset_version` is now exclusively available in `torch.onnx.OnnxRegistry` as it is removed from `torch.onnx.ExportOption`. The initialization of the registry with torchlib under the provided opset version ensures that the exporter uses the specified opset version as the primary version for exporting.

These changes encompass scenarios where users can:

1. Register an unsupported ATen operator with a custom implementation using onnx-script.
2. Override an existing symbolic function (onnx invariant).

NOTE: The custom registered function will be prioritized in onnx dispatcher, and if there are multiple custom ones, the one registered the last will be picked.

[ghstack-poisoned]

thiagocrepaldi commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -209,15 +209,29 @@ def forward(self, input):
-                return torch.ops.aten.add(input, input)
+                return torch.ops.aten.add.Tensor(input, input)
+
+        # _DEFAULT_OPSET_VERSION = 18
```
```suggestion
```
this comment would easily get outdated when onnxscript is updated in the future

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -75,14 +82,179 @@ class ONNXFakeContext:
 
 
+class OnnxRegistry:
+    """Registry for ONNX functions.
+
+    The registry maintains a mapping from qualified names to symbolic functions under a
+    fixed opset version. It supports registering custom onnx-script functions and for
+    dispatcher to dispatch calls to the appropriate function.
+
+    Attributes:
+        _registry: The registry maps OpNameto a list of SymbolicFunctions. It is important
+            not to directly modify this variable. Instead, access to it should be done through
+            the public methods: register_custom_op, get_functions, and is_registered_op.
+
```
```suggestion
    """Registry for ONNX operators.

    The registry maintains a mapping from PyTorch ATen operator qualified names to ONNX functions under a
    system-defined opset version. It supports registering custom operators.
```
Tried to NIT the docstring to hide implementation detail that the registry user shouldn't worry about.

Removed private stuff shouldn't be publicly documented. You can add a comment with this info somewhere, like where `self._registry` is initialized

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -75,14 +82,179 @@ class ONNXFakeContext:
+
+    def __init__(self) -> None:
+        """Initializes the registry"""
+        self._registry: Dict[
```
```suggestion

        # NOTE: _registry: The registry maps OpName to a list of SymbolicFunctions. It is important
        # not to directly modify this variable. Instead, access to it should be done through
        # the public methods.
        self._registry: Dict[
```

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -75,14 +82,179 @@ class ONNXFakeContext:
+        self._registry[internal_qualified_name].append(symbolic_function)
+
+    @_beartype.beartype
+    def register_custom_op(
```
```suggestion
    def register_op(
```
As discussed in during meeting, ideally this API could be used to register custom operators, but also to override existing ones, which would make the "custom" word confusing

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -75,14 +82,179 @@ class ONNXFakeContext:
+        overload: Optional[str] = None,
+        is_complex: bool = False,
+    ) -> None:
+        """Registers a custom operator: torch.ops.<namespace>.<op_name>.<overload>.
+
+        Args:
+            function: The onnx-sctip function to register.
+            namespace: The namespace of the operator to register.
+            op_name: The name of the operator to register.
+            overload: The overload of the operator to register. If it's default overload,
+                leave it to None.
+            is_complex: Whether the function is a function that handles complex valued inputs.
+
+        Raises:
+            ValueError: If the name is not in the form of 'namespace::op'.
```
```suggestion
        """Registers an operator with torch.ops.<namespace>.<op_name>.<overload> format.

        Args:
            function: The ONNX function to register.
            namespace: The namespace of the Aten operator to register.
            op_name: The name of the Aten operator to register.
            overload: The overload of the Aten operator to register. Set to None to mark as the default overload.
            is_complex: Whether the function is a function that handles complex valued inputs.

        Raises:
            ValueError: If the name is not in the form of 'namespace::op'.
```

titaiwangms replied:
Same as others

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -75,14 +82,179 @@ class ONNXFakeContext:
+        self._register(internal_name_instance, symbolic_function)
+
+    @_beartype.beartype
+    def get_functions(
```
```suggestion
    def get_op(
```

That is tricky. Although we want the onnx function under the hood, we registered that function using `register_op` (focus on `_op`), so fetching that function with `get_op` seems more intuitive.

users can add op, check whether the op is register and get the op. the function term seems secondary

titaiwangms replied:
Use `get_op_functions` instead. I think list of functions in return is still worths to mention.

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -75,14 +82,179 @@ class ONNXFakeContext:
+    def get_functions(
+        self, namespace: str, op_name: str, overload: Optional[str] = None
+    ) -> Optional[List[registration.SymbolicFunction]]:
+        """Returns a list of SymbolicFunctions for the given op: torch.ops.<namespace>.<op_name>.<overload>.
+
+        The list is ordered by the time of registration. The custom operators should be
+        in the second half of the list.
+
+        Args:
+            namespace: The namespace of the operator to get.
+            op_name: The name of the operator to get.
+            overload: The overload of the operator to get. If it's default overload,
+                leave it to None.
+        Returns:
+            A list of SymbolicFunctions corresponding to the given name, or None if
+            the name is not in the registry.
```
```suggestion
        """Returns a list of ONNX functions for torch.ops.<namespace>.<op_name>.<overload>.

        The list is prioritized according to the registration order, in a way that the most recently registered operator takes precedence.

        Args:
            namespace: The namespace of the Aten operator to get.
            op_name: The name of the Aten operator to get.
            overload: The overload of the Aten operator to get. Set to None to mark as the default overload.
        Returns:
            A list of ONNX Functions corresponding to the given name, or None if
            the name is not registered.
```

titaiwangms replied:
I will update the docstring wherever uses `SymbolicFunction` when I decide a better name.

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -75,14 +82,179 @@ class ONNXFakeContext:
+    def is_registered_op(
+        self, namespace: str, op_name: str, overload: Optional[str] = None
+    ) -> bool:
+        """Returns whether the given op is registered: torch.ops.<namespace>.<op_name>.<overload>.
+
+        Args:
+            namespace: The namespace of the operator to check.
+            op_name: The name of the operator to check.
+            overload: The overload of the operator to check. If it's default overload,
+                leave it to None.
```
```suggestion
        """Returns whether the given Aten torch.ops.<namespace>.<op_name>.<overload> is registered.

        Args:
            namespace: The namespace of the Aten operator to check.
            op_name: The name of the Aten operator to check.
            overload: The overload of the Aten operator to check. If it's default overload,
                leave it to None.
```

titaiwangms replied:
This part should remain. It could be other namespace.

thiagocrepaldi commented on torch/onnx/_internal/fx/registration.py:
```diff
@@ -14,9 +13,6 @@
-
-OpsetVersion = int
 
 
```
I would suggest renaming `SymbolicFunction` to something else. The `Symbolic` term is heavily overloaded lately, especially with dynamic shape (which has symbolic or static shape) and fake tensor (which has the idea of symbolic tracing)

Maybe something more neutral like `class ONNXFunction` or `OpImpl`, `ONNXScriptFunction` would be more meaningful for the user that will implement custom ops in the future

titaiwangms replied:
Good point

titaiwangms replied:
Go with `ONNXFunction`

titaiwangms committed:
Update on "[ONNX] Expose OnnxRegistry publicly"


The official move of `OnnxRegistry` to `torch.onnx` allows it to become one of the parameters in `torch.onnx.ExportOption`. By incorporating `OnnxRegistry` in `torch.onnx.ExportOption`, users gain access to various functionalities, including the ability to register custom operators using `register_custom_op`, check whether an operator is supported using `is_registered_op`, and obtain symbolic functions that support specific operators using `get_functions`.

Additionally, `opset_version` is now exclusively available in `torch.onnx.OnnxRegistry` as it is removed from `torch.onnx.ExportOption`. The initialization of the registry with torchlib under the provided opset version ensures that the exporter uses the specified opset version as the primary version for exporting.

These changes encompass scenarios where users can:

1. Register an unsupported ATen operator with a custom implementation using onnx-script.
2. Override an existing symbolic function (onnx invariant).

NOTE: The custom registered function will be prioritized in onnx dispatcher, and if there are multiple custom ones, the one registered the last will be picked.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Expose OnnxRegistry publicly"


The official move of `OnnxRegistry` to `torch.onnx` allows it to become one of the parameters in `torch.onnx.ExportOption`. By incorporating `OnnxRegistry` in `torch.onnx.ExportOption`, users gain access to various functionalities, including the ability to register custom operators using `register_custom_op`, check whether an operator is supported using `is_registered_op`, and obtain symbolic functions that support specific operators using `get_functions`.

Additionally, `opset_version` is now exclusively available in `torch.onnx.OnnxRegistry` as it is removed from `torch.onnx.ExportOption`. The initialization of the registry with torchlib under the provided opset version ensures that the exporter uses the specified opset version as the primary version for exporting.

These changes encompass scenarios where users can:

1. Register an unsupported ATen operator with a custom implementation using onnx-script.
2. Override an existing symbolic function (onnx invariant).

NOTE: The custom registered function will be prioritized in onnx dispatcher, and if there are multiple custom ones, the one registered the last will be picked.

[ghstack-poisoned]

titaiwangms commented:
I will have another PR to address, if there is any, other comments.
